### DISCUSSION TITLE
How do I protect against incorrect typing of a ForeignKey / relationship?

### DISCUSSION THREAD

**Discussion by Apakottur** [sqlalchemy/sqlalchemy#10148] [Usage Questions] [answered]
I've been testing out the new 2.0 declarative typing, documented [here](https://docs.sqlalchemy.org/en/20/changelog/whatsnew_20.html#whatsnew-20-orm-declarative-typing), and I cannot find how am I supposed to protect myself against typing a FK/relationship incorrectly.
Consider the following example, where I incorrectly annotate a FK/relationship:

```python
from sqlalchemy import ForeignKey
from sqlalchemy.orm import DeclarativeBase, mapped_column, relationship, Mapped

class Model(DeclarativeBase):
    pass

class Person(Model):
    id: Mapped[str] = mapped_column(primary_key=True)

class SomethingElse(Model):
    id: Mapped[str] = mapped_column(primary_key=True)

class Group(Model):
    # FK / relationship with incorrect typing:
    person_id: Mapped[int] = mapped_column(ForeignKey(Person.id))
    person: Mapped[SomethingElse] = relationship(Person)

person = Person(id="123")
group = Group()

reveal_type(person.id)  # Expected: str
reveal_type(group.person_id)  # Expected: str
reveal_type(group.person)  # Expected: Person
```

Running mypy without any flags gives:

```python
test.py:26:13: note: Revealed type is "builtins.str"
test.py:27:13: note: Revealed type is "builtins.int"
test.py:28:13: note: Revealed type is "test.SomethingElse"
```

Which matches my annotations, but not the correct types of these attributes. How do I protect myself against incorrectly annotating a FK/relationship?

When generating a DB from these models, the foreign key column type is set correctly to a string, but the type inferred by mypy is still an int.

**Comment by zzzeek** [comment, marked as answer]
well you wouldn't specify a class in the `relationship()`:

    person: Mapped[SomethingElse] = relationship(Person)

would be

    person: Mapped[SomethingElse] = relationship()

then it would fail at runtime due to lack of FKs

as far as types within `Mapped[]` otherwise not matching, SQLAlchemy does not try to runtime check these right now; it was extremely difficult to get interpretation of annotations to work at all, much less correctly, so there was a great emphasis on not getting involved with annotations at all if the right hand side had better information to work with, so that backwards compatibility with various approaches could be maintained as well as providing for workarounds for places where the left-hand side interpretation were not working correctly.

runtime reconciliation of the right side runtime configuration with the left side annotation as given could be added but this would be in the form of warnings only, and it's not clear what current workaround  / alternate approaches in use might be impacted by this.

**Reply by Apakottur** [reply]
Thanks for the answer!

Makes perfect sense, and getting a warning about a mismatch between the left and right sides would have been the perfect solution for my problem, I was just wondering if there was a way to get that kind of a warning that I was not aware of.

**Reply by zzzeek** [reply]
ideally typing alone could do it if you for example had `Mapped[int] = mapped_column(String)`, which...I think it does right now?  it should at least.  but if there's no type in `mapped_column()` then it's not possible with typing alone.

**Reply by zzzeek** [reply]
nah, it doesn't.  very tough to get typing to do all of this completely
